src/command_controller.rs

```diff
@@ -0,0 +1,288 @@
+//!     type Owner = Foo;
+//!
+//!     fn get_owner_name(&self) -> String {
+//!         self.spec.name.clone()
+//!     }
+//! }
+//!
+//! impl Crd for Bar {
+//!     const RESOURCE_NAME: &'static str = "bars.command.foo.stackable.tech";
+//!     const CRD_DEFINITION: &'static str = "
+//! apiVersion: apiextensions.k8s.io/v1
+//! kind: CustomResourceDefinition
+//! metadata:
+//!   name: bars.command.foo.stackable.tech
+//! spec:
+//!   group: command.foo.stackable.tech
+//!   names:
+//!     kind: Bar
+//!     singular: bar
+//!     plural: bars
+//!     listKind: BarList
+//!   scope: Namespaced
+//!   versions:
+//!     - name: v1
+//!       served: true
+//!       storage: true
+//!       schema:
+//!         openAPIV3Schema:
+//!           type: object
+//!           properties:
+//!             spec:
+//!               type: object
+//!               properties:
+//!                 name:
+//!                   type: string
+//!             status:
+//!               nullable: true
+//!               type: object
+//!               properties:
+//!                 startedAt:
+//!                   type: string
+//!                 finishedAt:
+//!                   type: string
+//!                 message:
+//!                   type: string
+//!       subresources:
+//!         status: { }";
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() -> Result<(), error::Error> {
+//!    stackable_operator::logging::initialize_logging("FOO_OPERATOR_LOG");
+//!    let client = client::create_client(Some("foo.stackable.tech".to_string())).await?;
+//!
+//!    stackable_operator::crd::ensure_crd_created::<Foo>(client.clone()).await?;
+//!    stackable_operator::crd::ensure_crd_created::<Bar>(client.clone()).await?;
+//!
+//!    tokio::join!(
+//!        stackable_foo_operator::create_controller(client.clone()),
+//!        stackable_operator::command_controller::create_command_controller::<Bar>(client)
+//!    );
+//!    Ok(())
+//! }
+//! ```
+//!
+use crate::client::Client;
+use crate::controller::{Controller, ControllerStrategy, ReconciliationState};
+use crate::error::{Error, OperatorResult};
+use crate::reconcile::{ReconcileFunctionAction, ReconcileResult, ReconciliationContext};
+use crate::{metadata, Crd};
+use async_trait::async_trait;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
+use kube::api::{ListParams, Meta};
+use kube::Api;
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+use std::future::Future;
+use std::pin::Pin;
+use std::time::Duration;
+
+const FINALIZER_NAME: &str = "command.stackable.tech/cleanup";
+
+/// The Owner trait represents our main controller object. This is required to be able to search
+/// for the main controller object and extract metadata information to set the owner reference
+/// in our command object. This must be implemented by every command custom resource.
+pub trait CommandOwner {
+    type Owner: Meta + Clone + DeserializeOwned + Debug + Send + Sync + Crd;
```

Comment:
We already talked about this offline: I believe we don't need this type in here.
It can be moved out and be another generic parameter for the methods below.